app/controllers/login_controller.rb

```diff
@@ -43,6 +50,37 @@ def refresh
     session[:oauth_response] = JSON.parse(response)
 
     # Redirect the user's browser to the intended home page
-    redirect_to users_home_path
+    redirect_to users_home_path, success: "You have successfully refreshed your access token!"
+
+    rescue RestClient::ExceptionWithResponse
+      if session[:oauth_response]
+        redirect_to users_home_path, danger: "Something Went Wrong. Please Refresh your Access Token and Try Again"
+      else
+        redirect_to login_index_path, danger: "Something Went Wrong. Please try again"
+      end
+    end
+
+  def revoke
+    # Populate the request body with the defined parameters from the docs
+    # https://developers.procore.com/reference/authentication#revoke-token
+    request = {
+      client_id: ENV['CLIENT_ID'],
+      client_secret: ENV['CLIENT_SECRET'],
+      token: session[:oauth_response]['access_token']
+    }
+    # Send a request to revoke the access token to Procore and return to the Sign In page.
+    # Note that in the production environment, this request uses the BASE_URL (https://api.procore.com) to revoke the token.
+    response = RestClient.post(ENV['BASE_URL'] + '/oauth/revoke', request.to_json)
+    # Deletes current session to clear out session variables
+    reset_session
+    # Redirects user back to the login page
+    redirect_to login_index_path
+
+    rescue RestClient::ExceptionWithResponse
+      if session[:oauth_response]
+        redirect_to users_home_path, danger: "Something Went Wrong. Please Refresh your Access Token and Try Again"
```

Comment:
My comments regarding the above error messages also apply to this one and the next one (on line 83).